### DISCUSSION TITLE
Springboot3.1.0 + skywalking + log4j2, but traceId is not print!

### DISCUSSION THREAD

**Discussion by zhangzx1996** [apache/skywalking#12415] [Q&A] [answered]
this is my log4j2.xml
`<?xml version="1.0" encoding="UTF-8"?>
<Configuration monitorInterval="10">

    <Properties>
        
        
        <property name="CONSOLE_PATTERN"
                  value="%d{yyyy-MM-dd HH:mm:ss.SSS} %-5level [%20.20t] [%30.30c{1.}]-%L T:[%traceId] S:[%X{spanId}]--%ecomsg%n"/>
        <property name="FILE_PATTERN"
                  value="%d{yyyy-MM-dd HH:mm:ss.SSS} %-5level [%20.20t] [%30.30c{1.}]-%L T:[%traceId] S:[%X{spanId}]--%ecomsg%n"/>
        
        <property name="FILE_PATH" value="export/Logs"/>

        <property name="ROLLING_FILE" value="rolling_file"/>
    </Properties>

    <appenders>
        <console name="Console" target="SYSTEM_OUT">
            
            <PatternLayout pattern="${CONSOLE_PATTERN}"/>
        </console>

        <RollingFile name="RollingInfoLog" fileName="${FILE_PATH}/info.log"
                     filePattern="${FILE_PATH}/${ROLLING_FILE}/info-%d{yyyy-MM-dd}-%i.log.gz">
            
            <ThresholdFilter level="DEBUG" onMatch="ACCEPT" onMismatch="DENY"/>
            <PatternLayout pattern="${FILE_PATTERN}"/>
            <Policies>
                
                <TimeBasedTriggeringPolicy interval="1"/>
                
                <SizeBasedTriggeringPolicy size="500MB"/>
            </Policies>
            
            <DefaultRolloverStrategy max="3"/>
        </RollingFile>

        
        <RollingFile name="RollingFileError" fileName="${FILE_PATH}/error.log"
                     filePattern="${FILE_PATH}/${ROLLING_FILE}/error-%d{yyyy-MM-dd}-%i.log.gz">
            
            <ThresholdFilter level="error" onMatch="ACCEPT" onMismatch="DENY"/>
            <PatternLayout pattern="${FILE_PATTERN}"/>
            <Policies>
                
                <TimeBasedTriggeringPolicy interval="1"/>
                
                <SizeBasedTriggeringPolicy size="500MB"/>
            </Policies>
            
            <DefaultRolloverStrategy max="3"/>
        </RollingFile>

        
        <RollingFile name="RollingDubboInfo" fileName="${FILE_PATH}/dubbo.log"
                     filePattern="${FILE_PATH}/${ROLLING_FILE}/dubbo-%d{yyyy-MM-dd}-%i.log.gz">
            
            <ThresholdFilter level="DEBUG" onMatch="ACCEPT" onMismatch="DENY"/>
            <PatternLayout pattern="${FILE_PATTERN}"/>
            <Policies>
                
                <TimeBasedTriggeringPolicy interval="1"/>
                
                <SizeBasedTriggeringPolicy size="500MB"/>
            </Policies>
            
            <DefaultRolloverStrategy max="3"/>
        </RollingFile>

    </appenders>

    <Loggers>
        
        <AsyncRoot includeLocation="false">
            
            <appender-ref ref="RollingFileError" level="ERROR"/>
            <appender-ref ref="Console"/>
            <appender-ref ref="RollingInfoLog"/>
        </AsyncRoot>

        
        <AsyncLogger name="org.apache.dubbo" additivity="false" includeLocation="false">
            <AppenderRef ref="RollingDubboInfo"/>
            <appender-ref ref="Console"/>
        </AsyncLogger>

    </Loggers>

</Configuration>`

**Comment by zhangzx1996** [comment]
I change AsyncRoot to Root will print traceId

**Reply by zhangzx1996** [reply]
But I want print log by async

**Comment by wu-sheng** [comment, marked as answer]
Use sync appender. SkyWalking toolkit has async mechanism internally.

**Reply by zhangzx1996** [reply]
Very thanks, I will use sync appender. I didn't found the SkyWalking toolkit async mechanis, can you give me a doc or something about this?

**Reply by wu-sheng** [reply]
It is in the codes, there is a memory queue for that.
